File: YuXiangROS/Chap2/2.4_2.5_Workspace/src/demo_cpp_pkg/src/writer_node.cpp
/*
1. 继承自父类
2. 调用父类方法
*/

#include "demo_cpp_pkg/person_node.hpp" // 必须新建头文件, 否则main函数会重复定义
#include <rclcpp/rclcpp.hpp>
#include <string>

class WriterNode : public PersonNode {
  private:
    std::string book_;

  public:
    WriterNode(const std::string& book) : PersonNode("writer_node", "ChangLi", 20) {
        this->book_ = book;
    }

    void write_book() {
        RCLCPP_INFO(this->get_logger(), "I am writing a book titled '%s'.", this->book_.c_str());
    }
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    auto writer_node = std::make_shared<WriterNode>("The Great Gatsby");
    writer_node->say_hello("reading");
    writer_node->write_book();
    rclcpp::spin(writer_node);
    rclcpp::shutdown();
    return 0;
}

File: YuXiangROS/Chap2/2.4_2.5_Workspace/src/demo_python_pkg/demo_python_pkg/writer_node.py
"""
1. 继承自父类
2. 调用父类方法
"""

import rclpy

# 必须用相对导入(demo_python_pkg.person_node或.person_node), 否则从系统库查找, 会报错
from demo_python_pkg.person_node import PersonNode


class WriterNode(PersonNode):
    def __init__(self, book: str):
        super().__init__("writer_node", "Changli", 20)  # 调用父类构造函数

        self.book = book

    def write_book(self):
        self.get_logger().info(f"I am writing a book titled '{self.book}'.")


def main(args=None):
    rclpy.init(args=args)
    writer_node = WriterNode("The Great Gatsby")
    writer_node.say_hello("reading")
    writer_node.write_book()
    try:
        rclpy.spin(writer_node)
    except KeyboardInterrupt:
        pass
    finally:
        writer_node.destroy_node()
        rclpy.shutdown()
